### PR TITLE
add possibility to set replica weights explicitly

### PR DESCRIPTION
**Pull request by Sovianum**
Add explicit replica weights support

### CODE REVIEW

**Review by g7r — CHANGES_REQUESTED** [review]

File: rediscluster/cluster.go [hunk 1]
```diff
@@ -98,14 +98,50 @@ type Opts struct {
 
 	// RoundRobinSeed - used to choose between master and replica.
 	RoundRobinSeed RoundRobinSeed
-	// LatencyOrientedRR - when MasterAndSlaves is used, prefer hosts with lower latency
+	// LatencyOrientedRR - when MasterAndSlaves is used, prefer hosts with lower latency (can not be set with non-empty WeightsByAddress)
 	LatencyOrientedRR bool
+	// WeightsByAddress - enables to explicitly set weights of replicas in all the shards; must include weights for all the replicas of all the shards (can not be set with LatencyOrientedRR)
+	WeightsByAddress map[string]uint32
```
[the remark this answers is not in the document]
Comment (Sovianum) [reply rediscluster/cluster.go:105]: Added fallback to default strategy in case of weight incompleteness

[the remark this answers is not in the document]
Comment (Sovianum) [reply rediscluster/cluster.go]: Switched to interface instead of map

**Review by g7r — CHANGES_REQUESTED** [review]

File: rediscluster/cluster.go [hunk 2]
```diff
@@ -64,6 +64,10 @@ const (
 	enabled  = 1
 )
 
+type WeightProvider interface {
+	GetWeightsByHost(hosts []string) ([]uint32, []bool)
```
[the remark this answers is not in the document]
Comment (Sovianum) [reply rediscluster/cluster.go]: Fixed

File: rediscluster/cluster.go [hunk 3]
```diff
@@ -64,6 +64,10 @@ const (
 	enabled  = 1
 )
 
+type WeightProvider interface {
```
[the remark this answers is not in the document]
Comment (Sovianum) [reply rediscluster/cluster.go:68]: Fixed

**Review by g7r — APPROVED** [review]

**Pull request merged**